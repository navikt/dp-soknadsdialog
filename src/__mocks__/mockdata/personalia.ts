import { IPersonalia } from "../../types/personalia.types";

export const mockPersonalia: IPersonalia = {
  folkeregistrertAdresse: {
    adresselinje1: "Brekkerødlia 41",
    adresselinje2: "",
    adresselinje3: "",
    postnummer: "1782",
    poststed: "Halden",
    landkode: "NO",
    land: "NORGE",
  },
  forNavn: "STORARTET",
  mellomNavn: "",
  etterNavn: "SOLSIKKE",
  fødselsDato: "1980-03-29",
  postAdresse: {
    adresselinje1: "Brekkerødlia 41",
    adresselinje2: "",
    adresselinje3: "",
    postnummer: "1782",
    poststed: "Halden",
    landkode: "NO",
    land: "NORGE",
  },
  ident: "29838099503",
  bankLandkode: "NO",
  banknavn: "Bank i utlandet",
  kontonummer: "12341212345",
};

export const mockPersonaliaStrentFortrolig: IPersonalia = {
  forNavn: "MEMORERENDE",
  mellomNavn: "",
  etterNavn: "PRODUKSJON",
  fødselsDato: "1928-01-15",
  kontonummer: "00825323708",
  ident: "15812849066",
};
