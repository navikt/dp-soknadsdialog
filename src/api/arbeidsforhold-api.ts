import { logRequestError } from "../error.logger";

export async function getArbeidsforhold(onBehalfOfToken: string): Promise<Response> {
  const url = `${process.env.API_BASE_URL}/arbeidsforhold`;
  const response = await fetch(url, {
    method: "GET",
    headers: {
      "Content-Type": "application/json",
      Accept: "application/json",
      Authorization: `Bearer ${onBehalfOfToken}`,
    },
  });

  if (!response.ok) {
    logRequestError(response.statusText, undefined, "Arbeidsforhold - Failed to get info");

    return new Response("Feil i uthenting av arbeidsforhold", {
      status: response.status,
    });
  }

  return response;
}
