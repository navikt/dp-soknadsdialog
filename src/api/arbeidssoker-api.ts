import { formatISO } from "date-fns";
import { GetServerSidePropsContext } from "next/types";
import { v4 as uuid } from "uuid";
import { getSession, getVeilarbregistreringOnBehalfOfToken } from "../utils/auth.utils";

export type IArbeidssokerStatus = "UNREGISTERED" | "REGISTERED" | "UNKNOWN";
export interface IArbeidssokerperioder {
  arbeidssokerperioder: [
    {
      fraOgMedDato: string;
      tilOgMedDato: string | null;
    }
  ];
}

export async function getArbeidssokerperioder({ req }: GetServerSidePropsContext) {
  const today = formatISO(new Date(), { representation: "date" });
  const session = await getSession(req);
  const onBehalfOfToken = await getVeilarbregistreringOnBehalfOfToken(session);

  const callId = uuid();
  const url = `${process.env.VEILARBPROXY_URL}/veilarbregistrering/api/arbeidssoker/perioder?fraOgMed=${today}`;

  return await fetch(url, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${onBehalfOfToken}`,
      "Nav-Consumer-Id": "dp-soknadsdialog",
      "Nav-Call-Id": callId,
    },
  });
}
