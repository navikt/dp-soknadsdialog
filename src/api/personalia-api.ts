import { logRequestError } from "../error.logger";

export async function getPersonalia(onBehalfOfToken: string) {
  const url = `${process.env.API_BASE_URL}/personalia`;
  const response = await fetch(url, {
    method: "GET",
    headers: {
      "Content-Type": "application/json",
      Accept: "application/json",
      Authorization: `Bearer ${onBehalfOfToken}`,
    },
  });

  if (!response.ok) {
    logRequestError(response.statusText, undefined, "Personalia - Failed to get info");

    return new Response("Feil i uthenting av gjeldende periode", {
      status: response.status,
    });
  }

  return response;
}
