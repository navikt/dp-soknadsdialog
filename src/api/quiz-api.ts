import { quizMalResponse } from "../localhost-data/quiz-mal-response";
import { formatISO, subDays } from "date-fns";
import { v4 as uuidV4 } from "uuid";

export const headersWithToken = (onBehalfOfToken: string) => ({
  "Content-Type": "application/json",
  Accept: "application/json",
  Authorization: `Bearer ${onBehalfOfToken}`,
});

export function getSoknadMal(onBehalfOfToken: string) {
  const url = `${process.env.API_BASE_URL}/soknad/mal`;

  if (process.env.NEXT_PUBLIC_LOCALHOST) {
    return Promise.resolve(quizMalResponse);
  }

  return fetch(url, {
    method: "GET",
    headers: headersWithToken(onBehalfOfToken),
  })
    .then((response: Response) => response.json())
    .catch((error) => {
      return Promise.reject(error);
    });
}

export function createSoknadUuid(onBehalfOfToken: string) {
  const url = `${process.env.API_BASE_URL}/soknad`;

  return fetch(url, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${onBehalfOfToken}`,
    },
  });
}

export function getMineSoknader(onBehalfOfToken: string) {
  const fromDate = subDays(Date.now(), 3);
  const formattedDate = formatISO(fromDate, { representation: "date" });

  const url = `${process.env.API_BASE_URL}/soknad/mine-soknader?fom=${formattedDate}`;
  return fetch(url, {
    method: "GET",
    headers: {
      "Content-Type": "application/json",
      Accept: "application/json",
      Authorization: `Bearer ${onBehalfOfToken}`,
    },
  });
}

export function createInnsendingUuid(onBehalfOfToken: string) {
  const url = `${process.env.API_BASE_URL}/soknad?prosesstype=Innsending`;

  return fetch(url, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${onBehalfOfToken}`,
    },
  });
}

export function getSoknadState(
  uuid: string,
  onBehalfOfToken: string,
  lastSaved?: string,
  requestId?: string
) {
  let url = `${process.env.API_BASE_URL}/soknad/${uuid}/neste`;

  if (lastSaved) {
    url += `?sistLagret=${lastSaved}`;
  }

  return fetch(url, {
    method: "GET",
    headers: {
      ...headersWithToken(onBehalfOfToken),
      "X-Request-ID": requestId || uuidV4(),
    },
  });
}

export function getSoknadStatus(uuid: string, onBehalfOfToken: string) {
  return fetch(`${process.env.API_BASE_URL}/soknad/${uuid}/status`, {
    method: "GET",
    headers: headersWithToken(onBehalfOfToken),
  });
}
