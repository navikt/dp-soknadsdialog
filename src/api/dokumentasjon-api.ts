import api from "../api.utils";

export async function saveDokumenkravFile(file: File, uuid: string, dokumentkravId: string) {
  const requestData = new FormData();
  requestData.append("file", file);

  // Do NOT specify content-type here, it gets browser generated with the correct boundary by default
  return fetch(api(`/documentation/${uuid}/${dokumentkravId}/file/save`), {
    method: "POST",
    headers: {
      accept: "application/json",
    },
    body: requestData,
  });
}
