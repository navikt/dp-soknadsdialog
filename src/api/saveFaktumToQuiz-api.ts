import api from "../api.utils";
import { QuizFaktum, QuizFaktumSvarType, IQuizGeneratorFaktum } from "../types/quiz.types";

export async function saveFaktum(
  uuid: string | string[] | undefined,
  faktum: QuizFaktum | IQuizGeneratorFaktum,
  svar: QuizFaktumSvarType | QuizFaktum[][]
) {
  const res = await fetch(api(`/soknad/${uuid}/faktum/${faktum.id}`), {
    method: "PUT",
    body: JSON.stringify({ ...faktum, svar }),
  });

  if (!res.ok) {
    throw new Error(res.statusText);
  }
  return res.json();
}
