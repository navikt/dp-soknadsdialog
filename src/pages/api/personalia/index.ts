import { NextApiHandler, NextApiRequest, NextApiResponse } from "next";
import { HttpProblem, Person } from "../../../types/personalia.types";
import { getSession } from "@navikt/dp-auth/server";
import { audience } from "../../../api.utils";

const isHttpProblem = (variableToCheck: unknown): variableToCheck is HttpProblem =>
  (variableToCheck as HttpProblem).title !== undefined;

export const personaliaHenter: NextApiHandler<Person | HttpProblem> = async (
  req: NextApiRequest,
  res: NextApiResponse<Person | HttpProblem>
) => {
  try {
    const { token, apiToken } = await getSession({ req });
    if (token && apiToken) {
      const onBehalfOfToken = await apiToken(audience);
      const person = await fetch(`${process.env.API_BASE_URL}/personalia`, {
        method: "Get",
        headers: {
          "Content-Type": "application/json",
          Accept: "application/json",
          Authorization: `Bearer ${onBehalfOfToken}`,
        },
      }).then(async (respons: Response) => {
        if (!respons.ok) {
          return Promise.reject<HttpProblem>(await respons.json());
        }
        return respons.json();
      });
      return res.json(person);
    } else {
      return res.status(401).json({
        status: 401,
        title: "Ikke innlogget",
        type: new URL("urn:oppslag:personalia"),
      });
    }
  } catch (error) {
    if (isHttpProblem(error)) {
      const httpProblem: HttpProblem = <HttpProblem>error;
      console.error(`Kall mot personalia API feilet. Feilmelding: ${httpProblem.title}`);
      return res
        .status(httpProblem.status !== undefined ? httpProblem.status : 500)
        .json(httpProblem);
    } else {
      const httpProblem: HttpProblem = {
        status: 500,
        title: "Noe galt ved personalia kall",
        type: new URL("urn:oppslag:personalia"),
      };
      console.error(`Kall mot personalia API feilet. Feilmelding: ${error}`);
      return res.status(500).json(httpProblem);
    }
  }
};
