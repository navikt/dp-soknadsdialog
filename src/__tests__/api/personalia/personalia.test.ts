import { createMocks } from "node-mocks-http";
import { HttpProblem, personaliaHandler } from "../../../pages/api/personalia";
import { getSession as _getSession } from "@navikt/dp-auth/server";
import { Personalia } from "../../../types/personalia.types";

jest.mock("@navikt/dp-auth/server");
const getSession = _getSession as jest.MockedFunction<typeof _getSession>;

global.fetch = jest.fn(() => {
  return Promise.resolve({
    headers: new Headers({ "content-type": "application/json" }),
    ok: true,
    json: () =>
      Promise.resolve<Personalia>({
        forNavn: "Donald",
        mellomNavn: "J",
        etterNavn: "Trumpf",
        fødselsDato: new Date("1940-06-14"),
        kontonummer: "12345677889",
        bankLandkode: "SWE",
        banknavn: "Eskilstuna Rekarne Sparbank",
        folkeregistrertAdresse: {
          adresselinje1: "Soloveien 232",
          adresselinje2: "C/O Plassen",
          adresselinje3: "Bortafor vegen",
          byEllerStedsnavn: "SOL",
          landkode: "SWE",
          land: "Sverige",
          postkode: "1121",
        },
        postAdresse: {
          adresselinje1: "Soloveien 232",
          adresselinje2: "C/O Plassen",
          adresselinje3: "Bortafor vegen",
          byEllerStedsnavn: "SOL",
          landkode: "SWE",
          land: "Sverige",
          postkode: "1121",
        },
      }),
  });
}) as jest.Mock;

beforeEach(() => {
  getSession.mockResolvedValue({
    token: "123",
    payload: { pid: "123123", exp: Date.now() / 1000 + 3000 },
    apiToken: async () => "foo",
  });
});
afterEach(() => {
  getSession.mockClear();
  (fetch as jest.Mock).mockClear();
});

describe("/api/personalia", () => {
  test("svarer med personalia", async () => {
    const { req, res } = createMocks({
      method: "GET",
    });

    // @ts-ignore MockRequest matcher ikke AuthedNextApiRequest
    await personaliaHandler(req, res);

    expect(res._getStatusCode()).toBe(200);
    expect(res._getJSONData()).toMatchSnapshot();
  });

  test("ved feil", async () => {
    // @ts-ignore Vet ikke hvordan en gjør det i ts
    fetch.mockImplementationOnce(() => Promise.reject("API is down"));
    const { req, res } = createMocks({
      method: "GET",
    });

    // @ts-ignore MockRequest matcher ikke AuthedNextApiRequest
    await personaliaHandler(req, res);

    expect(res._getStatusCode()).toBe(500);
  });

  test("APIet svarer med annen enn 200 ok", async () => {
    // @ts-ignore Vet ikke hvordan en gjør det i ts
    fetch.mockImplementationOnce(() =>
      Promise.resolve({
        headers: new Headers({ "content-type": "application/json" }),
        status: 400,
        ok: false,
        json: () =>
          Promise.resolve<HttpProblem>({
            detail: "En lengre beskrivelse av hvorfor det feiler",
            status: 400,
            title: "Feil token",
            type: new URL("urn:oppslag:personalia"),
          }),
      })
    );
    const { req, res } = createMocks({
      method: "GET",
    });

    // @ts-ignore MockRequest matcher ikke AuthedNextApiRequest
    await personaliaHandler(req, res);

    expect(res._getStatusCode()).toBe(400);
    expect(res._getJSONData()).toMatchSnapshot();
  });

  test("Uinnlogget", async () => {
    // @ts-ignore Vet ikke hvordan en gjør det i ts
    getSession.mockResolvedValue({});
    const { req, res } = createMocks({
      method: "GET",
    });

    // @ts-ignore MockRequest matcher ikke AuthedNextApiRequest
    await personaliaHandler(req, res);

    expect(res._getStatusCode()).toBe(401);
    expect(res._getJSONData()).toMatchSnapshot();
  });
});
