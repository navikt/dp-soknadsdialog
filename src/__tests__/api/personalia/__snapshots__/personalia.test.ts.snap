// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/api/personalia APIet svarer med annen enn 200 ok 1`] = `
Object {
  "detail": "En lengre beskrivelse av hvorfor det feiler",
  "status": 400,
  "title": "Feil token",
  "type": "urn:oppslag:personalia",
}
`;

exports[`/api/personalia Uinnlogget 1`] = `
Object {
  "status": 401,
  "title": "Ikke innlogget",
  "type": "urn:oppslag:personalia",
}
`;

exports[`/api/personalia svarer med personalia 1`] = `
Object {
  "etterNavn": "Trumpf",
  "folkeregistrertAdresse": Object {
    "adresselinje1": "Soloveien 232",
    "adresselinje2": "C/O Plassen",
    "adresselinje3": "Bortafor vegen",
    "byEllerStedsnavn": "SOL",
    "land": "Sverige",
    "landkode": "SWE",
    "postkode": "1121",
  },
  "forNavn": "Donald",
  "f√∏dselsDato": "1940-06-14T00:00:00.000Z",
  "mellomNavn": "J",
  "postAdresse": Object {
    "adresselinje1": "Soloveien 232",
    "adresselinje2": "C/O Plassen",
    "adresselinje3": "Bortafor vegen",
    "byEllerStedsnavn": "SOL",
    "land": "Sverige",
    "landkode": "SWE",
    "postkode": "1121",
  },
}
`;
